### PR TITLE
PLP module

### PR DESCRIPTION
**Pull request by zahinSJI**
Module complete
Report: [Link](https://drive.google.com/file/d/1GU65Lk3kEoMOtx8ky5WIa7q9-mzHum8B/view?usp=sharing)

### CODE REVIEW

**Comment by zahinSJI** [comment]
Here is the report with all modules: [Link](https://drive.google.com/file/d/1h2X4bN1Q7kZCKAqP4orGZvnFXt4bURz_/view?usp=sharing)

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    public ListingPage() {
+        PageFactory.initElements(driver, this);
+        actions = new Actions(driver);
+    }
+
+    /**
+     * Locators 
+     */
+    @FindBy(css = "#tab-menu-8651")
+    WebElement prodHeaderMenu;
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: @zahinSJI all locators should be private

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    @FindBy(css = "[data-value='56']")
+    WebElement adultOralQuickFilter;
+
+    @FindBy(css = "[data-value='711']")
+    WebElement adultTopicalQuickFilter;
+
+    @FindBy(css = "[data-value='61']")
+    WebElement kidsProdQuickFilter;
+	
+	@FindBy(xpath = "//div[@id='onetrust-close-btn-container']")
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java:53]: These locators are not aligned properly. Please format the source code
@zahinSJI

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	@FindBy(css = ".main-row .three-quarters")
+	WebElement productsPanel;
+	
+	@FindBy(id = "edit-sort-by")
+	WebElement sortBySelector;
+	
+	@FindBy(css = ".sfmc-careclub-lightbox-close")
+	private WebElement closeSignUpBtn;
+    
+    public WebElement getBanner(String text) {
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: This is returning a web locator. It should be private.
@zahinSJI

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        return driver.findElement(By.xpath("//div[@class='inner']/*[contains(.,'" + text + "')]"));
+    }
+    
+    public WebElement getFilter(String text) {
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: This is returning a web locator. It should be private.
@zahinSJI

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 1 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: All locator made private

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 2 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java:53]: Formatting done

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 3 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: All locators made private

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 4 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: All locators made private

Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: All locator made private

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 1 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: All locators made private

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 2 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java:53]: Formatting done

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 3 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: All locators made private

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+		Action.performActionwithExtentInfoLog(prodHeaderMenu, "click", "Clicking on : " + prodHeaderMenu.getText());
+	}
+
+	/**
+	 * Visits a listing page from the homepage. Then, verifies the banner image in
+	 * the listing page.
+	 */
+	public void verifyBannerImage() {
+		this.visitPLP();
+		extentInfoLog("Image is displayed : ", Action.isDisplayed(driver, bannerImage));
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: @zahinSJI where is the assertion ?

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+	/**
+	 * Visits a listing page from the homepage. Then, verifies the banner texts in
+	 * the listing page.
+	 */
+	public void verifyBannerTexts(String text) {
+		this.visitPLP();
+		WebElement actualElem = getBanner(text);
+		String actualText = actualElem.getText();
+		extentInfoLog("Banner heading is displayed : ", Action.isDisplayed(driver, actualElem));
+		Assert.assertEquals(actualText, text);
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: @zahinSJI use method printAndAssert() from Actions class

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+	/**
+	 * Visits a listing page from the homepage. Then, verifies the filters in the
+	 * listing page.
+	 */
+	public void verifyQuickFilters(String text) {
+		this.visitPLP();
+		WebElement actualElem = getFilter(text);
+		String actualText = actualElem.getText();
+		extentInfoLog("Filter text is displayed : ", Action.isDisplayed(driver, actualElem));
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: use proper assertions to check if element is displayed.
Assert.assertTrue(ele.isDsiplayed())

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+		Action.performActionwithExtentInfoLog(actualElem, "click", "Clicking on : " + actualText);
+		extentInfoLog("Products are displayed : ", Action.isDisplayed(driver, productsPanel));
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: use proper assertions to check if element is displayed.
Assert.assertTrue(ele.isDsiplayed())

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	/**
+	 * Visits a listing page from the homepage. Then, verifies the sort by selector
+	 * the listing page.
+	 */
+	public void verifySortBy(String text) {
+		this.visitPLP();
+		extentInfoLog("Sort by displayed : ", Action.isDisplayed(driver, sortBySelector));
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: use proper assertions to check if element is displayed.
Assert.assertTrue(ele.isDsiplayed())

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+		Action.selectByValue(sortBySelector, text);
+		extentInfoLog("Products are displayed : ", Action.isDisplayed(driver, productsPanel));
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: use proper assertions to check if element is displayed.
Assert.assertTrue(ele.isDsiplayed())

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	private WebElement getFilter(String text) {
+		return driver.findElement(By.xpath("//div[@class='filter-row-shortcuts']/a[contains(.,'" + text + "')]"));
+	}
+
+	/**
+	 * Function to click on close privacy pop-up button
+	 */
+	public void closePrivacyPopup() {
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: is this function used anywhere ? 
@zahinSJI

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+			extentPassLog("Privacy pop-up displayed : ", true);
+			Action.performActionwithExtentInfoLog(closePrivacyBtn, "click", "Clicking on : Close button");
+		} else
+			extentFailLog("Privacy pop-up displayed : ", false);
+	}
+
+	/**
+	 * Function to click on close sign up pop-up button
+	 */
+	public void closeSignupPopup() {
```
Comment (goutamnaik007) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Is this function used anywhere ? @zahinSJI

**Review by goutamnaik007 — CHANGES_REQUESTED** [review]
Please make changes as per code review
Also,
- Add TC's to verify pricespider is displayed for correct product after clicking on Buy Now
- Add TC's to verify article redirections
- Add TC's to verify ages filter
- Add TC's to verify symptoms filter
- Add TC's to verify category filter
- Also update TC names of existing TC's to give out proper information

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 5 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Assertion added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 7 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Assertion added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 11 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: This is used in the spec file

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 12 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: This function is no longer necessary and removed from the script

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 6 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Assertion added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 8 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Assertion added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 9 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Assertion added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 10 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Assertion added

**Comment by zahinSJI** [comment]
Mentioned test cases have been added. Here is the [test report](https://drive.google.com/file/d/1RAPebycF3g0oeUEFjiw8qdQZdCFFTPwZ/view?usp=sharing) for the entire site.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 1 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 2 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java:53]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 3 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 4 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 5 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 6 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 7 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 8 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 9 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 10 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Verified

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 11 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: I checked in the spec file for Listing. This function is called using the object 'home' and not 'plp' . can you please check if its really used ? and where ?

@zahinSJI

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java (hunk 12 above)
Comment (goutamnaik007) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ListingPage.java]: Checked

**Pull request merged**